lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -1255,9 +1252,16 @@ matchWitness(TypeChecker &tc,
   };
 
   // Match a type in the requirement to a type in the witness.
-  auto matchTypes = [&](Type reqType, Type witnessType) 
+  auto matchTypes = [&](Type reqType, Type witnessType)
                       -> Optional<RequirementMatch> {
-    cs->addConstraint(ConstraintKind::Equal, reqType, witnessType, locator);
+    ConstraintKind kind = ConstraintKind::Conversion;
+    if (auto func = dyn_cast<FuncDecl>(req)) {
+      auto name = req->getFullName().getBaseName();
+      if (func->isOperator() && name.str() == "==") {
+        kind = ConstraintKind::Equal;
+      }
+    }
+    cs->addConstraint(kind, witnessType, reqType, locator);
```

Comment:
The second example is correct. Passing a throwing function as a non-throwing function does not work, because the caller does not expect it to throw. If the function does throw (and we admitted this conversion) you would encounter undefined behavior.

Think of `(T) -> U` as a subtype of `(T) throws -> U`, but not the other way around.